core-rust/state-manager/src/mempool/mempool_manager.rs

```diff
@@ -166,36 +181,50 @@ impl MempoolManager {
             .collect()
     }
 
-    /// Checks the committability of a random subset of transactions and removes the rejected ones
-    /// from the mempool.
+    /// Updates the underlying [`PriorityMempool`] with all accumulated updates from [`self.deferred_updates_rx`].
+    /// Currently the only place that sends these updates is [`PendingTransactionResultCache`].
+    pub fn execute_deferred_updates(&self) {
+        let mut write_mempool = self.mempool.write();
+
+        let updates = self.deferred_updates_rx.lock();
+
+        // We use [`try_iter`] instead of [`iter`] because we don't want to wait since this is not called in it's own actor thread yet.
+        for mempool_update in updates.try_iter() {
+            match &mempool_update {
+                MempoolUpdate::StateVersionUpdate(StateVersionUpdate {
+                    payload_hash,
+                    state_version,
+                }) => {
+                    write_mempool
+                        .update_transaction_executed_state_version(payload_hash, *state_version);
+                }
+                MempoolUpdate::TransactionRejected(payload_hash) => {
+                    write_mempool.remove_by_payload_hash(payload_hash);
+                }
+            }
+        }
+    }
+
+    /// Checks the committability of a subset of transactions executed against earliest state versions
+    /// and removes the newly rejected ones from the mempool.
     /// Obeys the given limit on the number of actually executed (i.e. not cached) transactions.
     pub fn reevaluate_transaction_committability(&self, max_reevaluated_count: u32) {
-        // TODO: better selection of transactions based on last time/state version against it was reevaluated.
-        const MIN_TRANSACTIONS_TO_CHECK_FOR_REEVALUATION: u32 = 100;
-        let candidate_transactions = self.mempool.read().get_k_random_transactions(max(
-            max_reevaluated_count,
-            MIN_TRANSACTIONS_TO_CHECK_FOR_REEVALUATION,
-        )
-            as usize);
-
         let mut transactions_to_remove = Vec::new();
-        let mut reevaluated_count = 0;
-        for candidate_transaction in candidate_transactions {
-            let (record, was_cached) = self
+        for candidate_transaction in self
```

Comment:
_(note from author)_

Now we can directly start to reexecute transactions since all rejected transaction (hitting `PendingTransactionResultCache`) are removed via deferred updates in `execute_deferred_updates` called exactly before this.